src/graph/heterograph.cc

```diff
@@ -494,6 +500,39 @@ CompactGraphs(const std::vector<HeteroGraphPtr> &graphs) {
   return result;
 }
 
+constexpr uint64_t kDGLSerialize_HeteroGraph = 0xDD589FBE35224ABF;
+
+bool HeteroGraph::Load(dmlc::Stream* fs) {
+  uint64_t magicNum;
+  CHECK(fs->Read(&magicNum)) << "Invalid Magic Number";
+  CHECK_EQ(magicNum, kDGLSerialize_HeteroGraph) << "Invalid HeteroGraph Data";
+  auto meta_grptr = new ImmutableGraph(static_cast<COOPtr>(nullptr));
+  CHECK(fs->Read(meta_grptr)) << "Invalid Immutable Graph Data";
+  uint64_t num_relation_graphs;
+  CHECK(fs->Read(&num_relation_graphs)) << "Invalid num of relation graphs";
+  std::vector<HeteroGraphPtr> relgraphs;
+  for (size_t i = 0; i < num_relation_graphs; ++i) {
+    UnitGraph* ugptr = Serializer::EmptyUnitGraph();
+    CHECK(fs->Read(ugptr)) << "Invalid UnitGraph Data";
+    relgraphs.emplace_back(dynamic_cast<BaseHeteroGraph*>(ugptr));
```

Comment:
There is no need to up-cast a subclass pointer to its parent graph pointer.

Reply:
Manipulating raw pointers with shared pointer as result is dangerous. Why not the following?
```c++
std::vector<HeteroGraphPtr> relgraphs;
for (...) {
  UnitGraphPtr ugptr = Serializer::EmptyUnitGraph();
  CHECK(fs->Read(ugptr.get())) << ... ;
  relgraphs.push_back(ugptr);
}
```

Reply:
Also I really suggest implementing serializer handlers such as one [here](https://github.com/dmlc/dgl/blob/master/include/dgl/runtime/serializer.h). This automatically allows serializing type `std::vector<T>` where `T` has specialized `Handler<T>`.

Reply:
> Also I really suggest implementing serializer handlers such as one [here](https://github.com/dmlc/dgl/blob/master/include/dgl/runtime/serializer.h). This automatically allows serializing type `std::vector<T>` where `T` has specialized `Handler<T>`.

Actually it's the same as the current implementation using `Load/Save` with the template class below.
https://github.com/dmlc/dmlc-core/blob/649be18a8c55c48517861d67158a45dec54992ee/include/dmlc/serializer.h#L104